scale_build/image/update.py

```diff
@@ -0,0 +1,147 @@
+    run(['mksquashfs', UPDATE_DIR, UPDATE_FILE, '-noD'], logger=update_image_logger)
+    update_hash = run(['sha256sum', UPDATE_FILE]).stdout.decode(errors='ignore').strip()
+    with open(UPDATE_FILE_HASH, 'w') as f:
+        f.write(update_hash)
+
+    build_update_manifest(update_hash)
+
+
+def sign_manifest(signing_key, signing_pass):
+    run(
+        f'echo "{signing_pass}" | gpg -ab --batch --yes --no-use-agent --pinentry-mode loopback --passphrase-fd 0 '
+        f'--default-key {signing_key} --output {os.path.join(UPDATE_DIR, "MANIFEST.sig")} '
+        f'--sign {os.path.join(UPDATE_DIR, "MANIFEST")}', exception_msg='Failed gpg signing with SIGNING_PASSWORD',
+    )
+
+
+def install_rootfs_packages(update_image_logger):
+    os.makedirs(os.path.join(CHROOT_BASEDIR, 'etc/dpkg/dpkg.cfg.d'), exist_ok=True)
+    with open(os.path.join(CHROOT_BASEDIR, 'etc/dpkg/dpkg.cfg.d/force-unsafe-io'), 'w') as f:
+        f.write('force-unsafe-io')
+
+    run_in_chroot(['apt', 'update'], update_image_logger)
+
+    manifest = get_manifest()
+    for package in itertools.chain(
+        manifest['base-packages'], map(lambda d: d['package'], manifest['additional-packages'])
+    ):
+        run_in_chroot(['apt', 'install', '-V', '-y', package], update_image_logger)
+
+    # Do any custom rootfs setup
+    custom_rootfs_setup(update_image_logger)
+
+    # Do any pruning of rootfs
+    clean_rootfs(update_image_logger)
+
+    # Copy the default sources.list file
+    shutil.copy(CONF_SOURCES, os.path.join(CHROOT_BASEDIR, 'etc/apt/sources.list'))
+
+    run_in_chroot(['depmod'], update_image_logger, check=False)
+
+
+def custom_rootfs_setup(rootfs_logger):
+    # Any kind of custom mangling of the built rootfs image can exist here
+
+    # If we are upgrading a FreeBSD installation on USB, there won't be no opportunity to run truenas-initrd.py
+    # So we have to assume worse.
+    # If rootfs image is used in a Linux installation, initrd will be re-generated with proper configuration,
+    # so initrd we make now will only be used on the first boot after FreeBSD upgrade.
+    with open(os.path.join(CHROOT_BASEDIR, 'etc/default/zfs'), 'a') as f:
+        f.write('ZFS_INITRD_POST_MODPROBE_SLEEP=15')
+
+    run_in_chroot(['update-initramfs', '-k', 'all', '-u'], logger=rootfs_logger)
+
+    # Generate native systemd unit files for SysV services that lack ones to prevent systemd-sysv-generator warnings
+    tmp_systemd = os.path.join(CHROOT_BASEDIR, 'tmp/systemd')
+    os.makedirs(tmp_systemd)
+    run_in_chroot([
+        '/usr/lib/systemd/system-generators/systemd-sysv-generator', '/tmp/systemd', '/tmp/systemd', '/tmp/systemd'
+    ], rootfs_logger)
+    for unit_file in filter(lambda f: f.endswith('.service'), os.listdir(tmp_systemd)):
+        with open(os.path.join(tmp_systemd, unit_file), 'a') as f:
+            f.write(textwrap.dedent('''\
+                [Install]
+                WantedBy=multi-user.target
+            '''))
+
+    run(
```

Comment:
Can't this be done in python?

Reply:
I changed it as you advised the bash code was nicer ?

Reply:
( https://github.com/truenas/scale-build/pull/139/commits/13eb1435040eb479fd37a1c967f537dd0168bdf4#diff-bcc5f33a65d37e234ad4bd1e940546df9002f3f5b84efb2b01a99526695b9cc7L119 --- https://github.com/truenas/scale-build/pull/139#discussion_r617370994 )